### PR TITLE
use apn from db for s11 & handle the case when ue does not send out apn-master

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

**Comment by ghost** [comment]
retest this please

File: src/mme-app/handlers/attach_aia_ula.c [hunk 1]
```diff
@@ -136,6 +136,14 @@ process_ula_resp()
 	ue_entry->ambr.max_requested_bw_dl = ula_msg->max_requested_bw_dl;
 	ue_entry->ambr.max_requested_bw_ul = ula_msg->max_requested_bw_ul;
 
+	ue_entry->selected_apn.len = ula_msg->selected_apn.len;
+	log_msg(LOG_INFO, "APN length from ula msg is - %d\n",
+			ula_msg->selected_apn.len);
+	memcpy(ue_entry->selected_apn.val, ula_msg->selected_apn.val,
+			ula_msg->selected_apn.len);
+	log_msg(LOG_INFO, "APN name from ula msg is - %s\n",
```
Comment (thakurajayL) [inline src/mme-app/handlers/attach_aia_ula.c]: Print name & length in single LOG....

File: src/s11/handlers/create_session_handler.c [hunk 2]
```diff
@@ -251,10 +251,15 @@ create_session_processing()
 
 
 	ies[ieCount].type = IE_APN;
-	ies[ieCount].length = htons(g_csReqInfo->apn.len);
+	ies[ieCount].length = htons((int)(g_csReqInfo->selected_apn.len) + 1);
```
Comment (thakurajayL) [inline src/s11/handlers/create_session_handler.c]: typecasting is not required.

File: src/s11/handlers/create_session_handler.c [hunk 3]
```diff
@@ -251,10 +251,15 @@ create_session_processing()
...
+	log_msg(LOG_INFO, "g_csReqInfo->selected_apn.len = %d\n",
```
Comment (thakurajayL) [inline src/s11/handlers/create_session_handler.c]: You can remove this log now.

File: src/s11/handlers/create_session_handler.c [hunk 4]
```diff
@@ -251,10 +251,15 @@ create_session_processing()
...
+			(int)(g_csReqInfo->selected_apn.len));
 	ies[ieCount].instance = INSTANCE_ZERO;
-	memcpy(ies[ieCount].value, g_csReqInfo->apn.val,
-			g_csReqInfo->apn.len);
+	unsigned char apn_len[25] = {};
```
Comment (thakurajayL) [inline src/s11/handlers/create_session_handler.c:247]: Using unsigned char is good enough. No need to use array.
e.g. unsigned char app_len;
app_len = g_csReqInfo->selected_apn.len;

File: src/s1ap/handlers/attach_icsreq.c [hunk 5]
```diff
@@ -501,12 +501,26 @@ icsreq_processing()
...
-	buffer_copy(&g_ics_buffer, &datalen, sizeof(datalen));
-	buffer_copy(&g_ics_buffer, (char *)ies[3].esm_msg.apn.val, datalen);
-	
+	// There is one category of UE, they do not send not apn to MME.
+	// In this case, the apn length from esm will be 0.
+	// Then MME will use the selected apn name from HSS-DB.
+	if (ies[3].esm_msg.apn.len == 0 ) {
+		datalen = (int) g_icsReqInfo->selected_apn.len + 1;
+		buffer_copy(&g_ics_buffer, &datalen, sizeof(datalen));
+		unsigned char selected_apn_len[25] = {};
```
Comment (thakurajayL) [inline src/s1ap/handlers/attach_icsreq.c:501]: no need to have array.

**Review by thakurajayL — CHANGES_REQUESTED** [review]
Can you please update code.

**Comment by ghost** [comment]
sure, thanks

File: src/s11/handlers/create_session_handler.c (hunk 3 above)
Comment (ghost) [reply src/s11/handlers/create_session_handler.c]: > Can you please update code.

sure, thanks

File: src/s11/handlers/create_session_handler.c (hunk 4 above)
Comment (ghost) [reply src/s11/handlers/create_session_handler.c:247]: the apn_len it self is no need an array.
later on it is used in strcat(apn_len, g_csReqInfo->selected_apn.val), this is why make it as array, the name is a confusing. I will change it.

File: src/mme-app/handlers/attach_aia_ula.c (hunk 1 above)
Comment (ghost) [reply src/mme-app/handlers/attach_aia_ula.c]: done

File: src/s11/handlers/create_session_handler.c (hunk 2 above)
Comment (ghost) [reply src/s11/handlers/create_session_handler.c]: done

File: src/s1ap/handlers/attach_icsreq.c (hunk 5 above)
Comment (ghost) [reply src/s1ap/handlers/attach_icsreq.c:501]: fixed in paging branch

File: src/s11/handlers/create_session_handler.c (hunk 4 above)
Comment (ghost) [reply src/s11/handlers/create_session_handler.c:247]: fixed in paging branch.

**Review by thakurajayL — APPROVED** [review]

**Pull request merged**